Source/WebKit/WebProcess/WebPage/WebPage.cpp

```diff
@@ -5366,24 +5368,65 @@ void WebPage::performDragControllerAction(std::optional<FrameIdentifier> frameID
     ASSERT_NOT_REACHED();
 }
 
-void WebPage::performDragOperation(WebCore::DragData&& dragData, SandboxExtension::Handle&& sandboxExtensionHandle, Vector<SandboxExtension::Handle>&& sandboxExtensionsHandleArray, CompletionHandler<void(bool)>&& completionHandler)
+static std::optional<WebCore::RemoteUserInputEventData> remoteEventDataForPosition(LocalFrame& frame, const IntPoint& position)
+{
+    auto locationInContentCoordinates = frame.view()->rootViewToContents(position);
+    auto hitTestResult = frame.eventHandler().hitTestResultAtPoint(locationInContentCoordinates, {
+        HitTestRequest::Type::ReadOnly,
+        HitTestRequest::Type::Active,
+        HitTestRequest::Type::DisallowUserAgentShadowContentExceptForImageOverlays,
+        HitTestRequest::Type::AllowChildFrameContent,
+    });
+    auto subframe = EventHandler::subframeForTargetNode(hitTestResult.protectedTargetNode().get());
+    RefPtr remoteFrame = dynamicDowncast<RemoteFrame>(subframe).get();
+    if (!remoteFrame)
+        return std::nullopt;
+    RefPtr remoteFrameView = remoteFrame->view();
+    if (!remoteFrameView)
+        return std::nullopt;
+    return RemoteUserInputEventData {
+        remoteFrame->frameID(),
+        remoteFrameView->rootViewToContents(position),
+    };
+}
+
+void WebPage::performDragOperation(std::optional<WebCore::FrameIdentifier> frameID, WebCore::DragData&& dragData, SandboxExtension::Handle&& sandboxExtensionHandle, Vector<SandboxExtension::Handle>&& sandboxExtensionsHandleArray, CompletionHandler<void(bool, std::optional<WebCore::RemoteUserInputEventData>, std::optional<SandboxExtensionHandle>, std::optional<Vector<SandboxExtension::Handle>>)>&& completionHandler)
 {
     ASSERT(!m_pendingDropSandboxExtension);
 
+    auto* frame = frameID ? WebProcess::singleton().webFrame(*frameID) : &mainWebFrame();
+
+    if (!frame) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    RefPtr localFrame = frame->coreLocalFrame();
+    if (!localFrame) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    auto remoteUserInputEventData = remoteEventDataForPosition(*localFrame, dragData.clientPosition());
```

Comment:
I think this will unnecessarily hit test twice, even when site isolation is disabled. Maybe instead we could reuse the hit test that happens in `DragController::performDragOperation`, and return `RemoteUserInputEventData` from that function. That would be similar to what we do in `DragController::dragEnteredOrUpdated`.